File: app/src/main/java/ohi/andre/consolelauncher/commands/tuixt/TuixtActivity.java
package ohi.andre.consolelauncher.commands.tuixt;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.text.TextUtils;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.Gravity;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.FrameLayout;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import java.io.File;
import java.util.ArrayList;
import java.util.List;

import ohi.andre.consolelauncher.managers.settings.LauncherSettings;
import ohi.andre.consolelauncher.managers.xml.XMLPrefsManager;
import ohi.andre.consolelauncher.managers.xml.classes.XMLPrefsSave;
import ohi.andre.consolelauncher.managers.xml.options.Behavior;
import ohi.andre.consolelauncher.tuils.LauncherSystemUi;

public class TuixtActivity extends Activity {

    public static final String PATH = "path";
    public static final String ERROR_KEY = "error";
    public static final int BACK_PRESSED = 2;
    public static final int SAVE_PRESSED = 3;

    private File file;
    private RecyclerView recyclerView;
    private TuixtAdapter adapter;
    private XMLPrefsManager.XMLPrefsRoot xmlRoot;
    private List<XMLPrefsSave> originalItems;
    private EditText plainTextEditor;
    private String originalRawText;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        LauncherSystemUi.requestNoTitleIfFullscreen(this);
        super.onCreate(savedInstanceState);
        LauncherSystemUi.applyFullscreen(this);

        Intent intent = getIntent();
        String path = intent.getStringExtra(PATH);
        if (path == null) {
            finish();
            return;
        }
        file = new File(path);

        FrameLayout screen = new FrameLayout(this);
        screen.setBackgroundColor(TuixtTheme.overlayColor());
        screen.setFitsSystemWindows(true);
        FrameLayout contentHost = TuixtLayout.addFoldAwareHost(this, screen, ViewGroup.LayoutParams.MATCH_PARENT);

        LinearLayout root = new LinearLayout(this);
        root.setOrientation(LinearLayout.VERTICAL);
        root.setPadding(TuixtTheme.dp(this, 14), TuixtTheme.dp(this, 50), TuixtTheme.dp(this, 14), TuixtTheme.dp(this, 14));
        TuixtTheme.stylePanel(this, root);

        int panelLeft = TuixtTheme.dp(this, 28);
        int panelTop = TuixtTheme.dp(this, 34);
        FrameLayout.LayoutParams panelParams = new FrameLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.MATCH_PARENT);
        panelParams.setMargins(panelLeft, panelTop, TuixtTheme.dp(this, 28), TuixtTheme.dp(this, 28));
        contentHost.addView(root, panelParams);

        TextView header = new TextView(this);
        header.setText("Themer/ " + file.getName());
        TuixtTheme.styleHeader(this, header);
        FrameLayout.LayoutParams headerParams = new FrameLayout.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT);
        headerParams.gravity = Gravity.TOP | Gravity.START;
        headerParams.leftMargin = panelLeft + TuixtTheme.dp(this, 38);
        headerParams.topMargin = panelTop - TuixtTheme.dp(this, 11);
        contentHost.addView(header, headerParams);

        // RecyclerView
        recyclerView = new RecyclerView(this);
        recyclerView.setLayoutParams(new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, 0, 1));
        recyclerView.setLayoutManager(new LinearLayoutManager(this));
        root.addView(recyclerView);

        // Bottom Bar (Search + Buttons)
        LinearLayout bottomBar = new LinearLayout(this);
        bottomBar.setOrientation(LinearLayout.VERTICAL);
        bottomBar.setPadding(10, 10, 10, 10);
        bottomBar.setBackground(TuixtTheme.rect(this, TuixtTheme.surfaceColor(), TuixtTheme.borderColor(), 1.25f));

        // Search Box
        EditText searchBox = new EditText(this);
        searchBox.setHint("Search settings...");
        TuixtTheme.styleInput(this, searchBox);
        bottomBar.addView(searchBox);

        // Action Buttons
        LinearLayout btnLayout = new LinearLayout(this);
        btnLayout.setOrientation(LinearLayout.HORIZONTAL);
        btnLayout.setPadding(0, 10, 0, 0);

        TextView btnCancel = new TextView(this);
        btnCancel.setText("CANCEL");
        TuixtTheme.styleButton(this, btnCancel, false);
        btnCancel.setOnClickListener(v -> attemptClose());
        btnLayout.addView(btnCancel);

        View spacer = new View(this);
        spacer.setLayoutParams(new LinearLayout.LayoutParams(0, 1, 1));
        btnLayout.addView(spacer);

        TextView btnSave = new TextView(this);
        btnSave.setText("SAVE");
        TuixtTheme.styleButton(this, btnSave, true);
        btnSave.setOnClickListener(v -> {
            Toast.makeText(this, "Applying changes...", Toast.LENGTH_SHORT).show();
            if (adapter != null) {
                adapter.saveAll(this);
            } else if (plainTextEditor != null) {
                try {
                    ohi.andre.consolelauncher.tuils.Tuils.write(file, "", plainTextEditor.getText().toString());
                    XMLPrefsManager.dispose();
                    XMLPrefsManager.loadCommons(this);
                    LauncherSettings.refreshFromLoadedPrefs();
                } catch (Exception e) {
                    Toast.makeText(this, "Error saving: " + e.getMessage(), Toast.LENGTH_LONG).show();
                }
            }
            setResult(SAVE_PRESSED);
            finish();
        });
        btnLayout.addView(btnSave);

        bottomBar.addView(btnLayout);
        root.addView(bottomBar);

        // Load data
        String fileName = file.getName().toLowerCase();
        for (XMLPrefsManager.XMLPrefsRoot rootEnum : XMLPrefsManager.XMLPrefsRoot.values()) {
            if (fileName.equals(rootEnum.path())) {
                xmlRoot = rootEnum;
                break;
            }
        }

        if (xmlRoot != null) {
            originalItems = new ArrayList<>(xmlRoot.enums);
            originalItems.remove(Behavior.toggle_output_state);
            adapter = new TuixtAdapter(originalItems, file);
            recyclerView.setAdapter(adapter);

            searchBox.addTextChangedListener(new TextWatcher() {
                @Override public void beforeTextChanged(CharSequence s, int start, int count, int after) {}
                @Override public void onTextChanged(CharSequence s, int start, int before, int count) {
                    filter(s.toString());
                }
                @Override public void afterTextChanged(Editable s) {}
            });
        } else {
            recyclerView.setVisibility(View.GONE);
            searchBox.setVisibility(View.GONE);

            plainTextEditor = new EditText(this);
            plainTextEditor.setLayoutParams(new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, 0, 1));
            plainTextEditor.setGravity(android.view.Gravity.TOP);
            TuixtTheme.styleInput(this, plainTextEditor);

            try {
                java.io.FileInputStream fis = new java.io.FileInputStream(file);
                originalRawText = ohi.andre.consolelauncher.tuils.Tuils.convertStreamToString(fis);
                plainTextEditor.setText(originalRawText);
                fis.close();
            } catch (Exception e) {
                originalRawText = "";
                plainTextEditor.setText("");
            }

            root.addView(plainTextEditor, 2);
        }

        setContentView(screen);
    }

    @Override
    protected void onResume() {
        super.onResume();
        LauncherSystemUi.applyFullscreen(this);
    }

    @Override
    public void onWindowFocusChanged(boolean hasFocus) {
        super.onWindowFocusChanged(hasFocus);
        if (hasFocus) {
            LauncherSystemUi.applyFullscreen(this);
        }
    }

    private void filter(String query) {
        List<XMLPrefsSave> filtered = new ArrayList<>();
        for (XMLPrefsSave item : originalItems) {
            if (item.label().toLowerCase().contains(query.toLowerCase()) || 
                item.info().toLowerCase().contains(query.toLowerCase())) {
                filtered.add(item);
            }
        }
        adapter.updateList(filtered);
    }

    @Override
    public void onBackPressed() {
        attemptClose();
    }

    private void attemptClose() {
        if (!hasUnsavedChanges()) {
            setResult(BACK_PRESSED);
            finish();
            return;
        }

        TuixtDialog.showConfirm(
                this,
                "Discard Changes?",
                "Unsaved settings changes will be lost.",
                "Discard",
                "Keep Editing",
                () -> {
                    setResult(BACK_PRESSED);
                    finish();
                });
    }

    private boolean hasUnsavedChanges() {
        if (adapter != null && adapter.hasPendingChanges()) {
            return true;
        }
        if (plainTextEditor != null) {
            return !TextUtils.equals(originalRawText, plainTextEditor.getText().toString());
        }
        return false;
    }
}

File: app/src/main/java/ohi/andre/consolelauncher/commands/tuixt/TuixtActivity.kt
package ohi.andre.consolelauncher.commands.tuixt

import android.annotation.SuppressLint
import android.app.Activity
import android.os.Bundle
import android.text.Editable
import android.text.TextUtils
import android.text.TextWatcher
import android.view.Gravity
import android.view.View
import android.view.ViewGroup
import android.widget.EditText
import android.widget.FrameLayout
import android.widget.LinearLayout
import android.widget.TextView
import android.widget.Toast
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import ohi.andre.consolelauncher.commands.tuixt.TuixtDialog.ConfirmAction
import ohi.andre.consolelauncher.commands.tuixt.TuixtLayout.addFoldAwareHost
import ohi.andre.consolelauncher.commands.tuixt.TuixtTheme.borderColor
import ohi.andre.consolelauncher.commands.tuixt.TuixtTheme.dp
import ohi.andre.consolelauncher.commands.tuixt.TuixtTheme.overlayColor
import ohi.andre.consolelauncher.commands.tuixt.TuixtTheme.rect
import ohi.andre.consolelauncher.commands.tuixt.TuixtTheme.styleButton
import ohi.andre.consolelauncher.commands.tuixt.TuixtTheme.styleHeader
import ohi.andre.consolelauncher.commands.tuixt.TuixtTheme.styleInput
import ohi.andre.consolelauncher.commands.tuixt.TuixtTheme.stylePanel
import ohi.andre.consolelauncher.commands.tuixt.TuixtTheme.surfaceColor
import ohi.andre.consolelauncher.managers.settings.LauncherSettings.refreshFromLoadedPrefs
import ohi.andre.consolelauncher.managers.xml.XMLPrefsManager
import ohi.andre.consolelauncher.managers.xml.XMLPrefsManager.XMLPrefsRoot
import ohi.andre.consolelauncher.managers.xml.classes.XMLPrefsSave
import ohi.andre.consolelauncher.managers.xml.options.Behavior
import ohi.andre.consolelauncher.tuils.LauncherSystemUi.applyFullscreen
import ohi.andre.consolelauncher.tuils.LauncherSystemUi.requestNoTitleIfFullscreen
import ohi.andre.consolelauncher.tuils.Tuils
import java.io.File
import java.io.FileInputStream
import java.util.Locale
import android.content.Intent
import java.util.ArrayList
import ohi.andre.consolelauncher.managers.settings.LauncherSettings
import ohi.andre.consolelauncher.tuils.LauncherSystemUi
import javax.xml.parsers.DocumentBuilderFactory
import org.w3c.dom.Element
import org.w3c.dom.Node

class TuixtActivity : Activity() {
    private var file: File? = null
    private var recyclerView: RecyclerView? = null
    private var adapter: TuixtAdapter? = null
    private var xmlRoot: XMLPrefsRoot? = null
    private var originalRows: MutableList<TuixtAdapter.SettingsRow>? = null
    private var plainTextEditor: EditText? = null
    private var originalRawText: String? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        requestNoTitleIfFullscreen(this)
        super.onCreate(savedInstanceState)
        applyFullscreen(this)

        val intent = getIntent()
        val path = intent.getStringExtra(PATH)
        if (path == null) {
            finish()
            return
        }
        file = File(path)

        val screen = FrameLayout(this)
        screen.setBackgroundColor(overlayColor())
        screen.setFitsSystemWindows(true)
        val contentHost = addFoldAwareHost(this, screen, ViewGroup.LayoutParams.MATCH_PARENT)

        val root = LinearLayout(this)
        root.setOrientation(LinearLayout.VERTICAL)
        root.setPadding(dp(this, 14f), dp(this, 50f), dp(this, 14f), dp(this, 14f))
        stylePanel(this, root)

        val panelLeft = dp(this, 28f)
        val panelTop = dp(this, 34f)
        val panelParams = FrameLayout.LayoutParams(
            ViewGroup.LayoutParams.MATCH_PARENT,
            ViewGroup.LayoutParams.MATCH_PARENT
        )
        panelParams.setMargins(panelLeft, panelTop, dp(this, 28f), dp(this, 28f))
        contentHost.addView(root, panelParams)

        val header = TextView(this)
        header.setText("Themer/ " + file!!.getName())
        styleHeader(this, header)
        val headerParams = FrameLayout.LayoutParams(
            ViewGroup.LayoutParams.WRAP_CONTENT,
            ViewGroup.LayoutParams.WRAP_CONTENT
        )
        headerParams.gravity = Gravity.TOP or Gravity.START
        headerParams.leftMargin = panelLeft + dp(this, 38f)
        headerParams.topMargin = panelTop - dp(this, 11f)
        contentHost.addView(header, headerParams)

        // RecyclerView
        recyclerView = RecyclerView(this)
        recyclerView!!.setLayoutParams(
            LinearLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                0,
                1f
            )
        )
        recyclerView!!.setLayoutManager(LinearLayoutManager(this))
        root.addView(recyclerView)

        // Bottom Bar (Search + Buttons)
        val bottomBar = LinearLayout(this)
        bottomBar.setOrientation(LinearLayout.VERTICAL)
        bottomBar.setPadding(10, 10, 10, 10)
        bottomBar.setBackground(rect(this, surfaceColor(), borderColor(), 1.25f))

        // Search Box
        val searchBox = EditText(this)
        searchBox.setHint("Search settings...")
        styleInput(this, searchBox)
        bottomBar.addView(searchBox)

        // Action Buttons
        val btnLayout = LinearLayout(this)
        btnLayout.setOrientation(LinearLayout.HORIZONTAL)
        btnLayout.setPadding(0, 10, 0, 0)

        val btnCancel = TextView(this)
        btnCancel.setText("CANCEL")
        styleButton(this, btnCancel, false)
        btnCancel.setOnClickListener(View.OnClickListener { v: View? -> attemptClose() })
        btnLayout.addView(btnCancel)

        val spacer = View(this)
        spacer.setLayoutParams(LinearLayout.LayoutParams(0, 1, 1f))
        btnLayout.addView(spacer)

        val btnSave = TextView(this)
        btnSave.setText("SAVE")
        styleButton(this, btnSave, true)
        btnSave.setOnClickListener(View.OnClickListener { v: View? ->
            Toast.makeText(this, "Applying changes...", Toast.LENGTH_SHORT).show()
            if (adapter != null) {
                adapter!!.saveAll(this)
            } else if (plainTextEditor != null) {
                try {
                    Tuils.write(file, "", plainTextEditor!!.getText().toString())
                    XMLPrefsManager.dispose()
                    XMLPrefsManager.loadCommons(this)
                    refreshFromLoadedPrefs()
                } catch (e: Exception) {
                    Toast.makeText(this, "Error saving: " + e.message, Toast.LENGTH_LONG).show()
                }
            }
            setResult(SAVE_PRESSED)
            finish()
        })
        btnLayout.addView(btnSave)

        bottomBar.addView(btnLayout)
        root.addView(bottomBar)

        // Load data
        val fileName = file!!.getName().lowercase(Locale.getDefault())
        for (rootEnum in XMLPrefsRoot.entries.toTypedArray()) {
            if (fileName == rootEnum.path()) {
                xmlRoot = rootEnum
                break
            }
        }

        if (xmlRoot != null) {
            originalRows = buildRows(xmlRoot!!, file!!)
            adapter = TuixtAdapter(originalRows!!.toMutableList(), file)
            recyclerView!!.setAdapter(adapter)

            searchBox.addTextChangedListener(object : TextWatcher {
                override fun beforeTextChanged(
                    s: CharSequence?,
                    start: Int,
                    count: Int,
                    after: Int
                ) {
                }

                override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                    filter(s.toString())
                }

                override fun afterTextChanged(s: Editable?) {}
            })
        } else {
            recyclerView!!.setVisibility(View.GONE)
            searchBox.setVisibility(View.GONE)

            plainTextEditor = EditText(this)
            plainTextEditor!!.setLayoutParams(
                LinearLayout.LayoutParams(
                    ViewGroup.LayoutParams.MATCH_PARENT,
                    0,
                    1f
                )
            )
            plainTextEditor!!.setGravity(Gravity.TOP)
            TuixtTheme.styleInput(this, plainTextEditor!!)

            try {
                val fis = FileInputStream(file)
                originalRawText = Tuils.convertStreamToString(fis)
                plainTextEditor!!.setText(originalRawText)
                fis.close()
            } catch (e: Exception) {
                originalRawText = ""
                plainTextEditor!!.setText("")
            }

            root.addView(plainTextEditor, 2)
        }

        setContentView(screen)
    }

    override fun onResume() {
        super.onResume()
        applyFullscreen(this)
    }

    override fun onWindowFocusChanged(hasFocus: Boolean) {
        super.onWindowFocusChanged(hasFocus)
        if (hasFocus) {
            applyFullscreen(this)
        }
    }

    private fun filter(query: String) {
        val rows = originalRows ?: return
        if (query.trim().isEmpty()) {
            adapter!!.updateRows(rows.toMutableList())
            return
        }

        val filtered: MutableList<TuixtAdapter.SettingsRow> = ArrayList()
        var pendingSection: TuixtAdapter.SettingsRow? = null
        val lower = query.lowercase(Locale.getDefault())
        for (row in rows) {
            if (row.sectionHeader) {
                pendingSection = row
                continue
            }

            val item = row.item ?: continue
            val label = item.label()!!.lowercase(Locale.getDefault())
            val info = item.info()!!.lowercase(Locale.getDefault())
            if (label.contains(lower) || info.contains(lower)) {
                if (pendingSection != null && (filtered.isEmpty() || filtered.last().section != pendingSection.section)) {
                    filtered.add(pendingSection)
                }
                filtered.add(row)
            }
        }
        adapter!!.updateRows(filtered)
    }

    private fun buildRows(root: XMLPrefsRoot, source: File): MutableList<TuixtAdapter.SettingsRow> {
        val remaining: LinkedHashMap<String, XMLPrefsSave> = LinkedHashMap<String, XMLPrefsSave>()
        for (save in root.enums) {
            if (save === Behavior.toggle_output_state) {
                continue
            }
            remaining[save.label()!!] = save
        }

        val groupedRows: LinkedHashMap<String, MutableList<XMLPrefsSave>> =
            LinkedHashMap<String, MutableList<XMLPrefsSave>>()
        try {
            val doc = DocumentBuilderFactory.newInstance().newDocumentBuilder().parse(source)
            val xmlRoot = doc.documentElement
            val children = xmlRoot.childNodes
            if (!hasSectionComments(children)) {
                val rows: MutableList<TuixtAdapter.SettingsRow> = ArrayList()
                addDefaultSectionRows(rows, remaining)
                return rows
            }

            var activeSection: String? = null
            for (index in 0..<children.length) {
                val node = children.item(index)
                if (node.nodeType == Node.COMMENT_NODE) {
                    val section = parseSectionComment(node.nodeValue)
                    if (section != null) {
                        activeSection = section
                    }
                } else if (node.nodeType == Node.ELEMENT_NODE) {
                    val element = node as Element
                    val save = remaining.remove(element.nodeName) ?: continue
                    val section = activeSection ?: XMLPrefsManager.sectionFor(save)
                    addGroupedRow(groupedRows, section, save)
                }
            }
        } catch (ignored: Exception) {
        }

        for (save in remaining.values) {
            val section = XMLPrefsManager.sectionFor(save)
            addGroupedRow(groupedRows, section, save)
        }

        return flattenGroupedRows(groupedRows)
    }

    private fun addDefaultSectionRows(
        rows: MutableList<TuixtAdapter.SettingsRow>,
        remaining: LinkedHashMap<String, XMLPrefsSave>
    ) {
        val groupedRows: LinkedHashMap<String, MutableList<XMLPrefsSave>> =
            LinkedHashMap<String, MutableList<XMLPrefsSave>>()
        val iterator = remaining.entries.iterator()
        while (iterator.hasNext()) {
            val save = iterator.next().value
            val section = XMLPrefsManager.sectionFor(save)
            addGroupedRow(groupedRows, section, save)
            iterator.remove()
        }
        rows.addAll(flattenGroupedRows(groupedRows))
    }

    private fun hasSectionComments(children: org.w3c.dom.NodeList): Boolean {
        for (index in 0..<children.length) {
            val node = children.item(index)
            if (node.nodeType == Node.COMMENT_NODE && parseSectionComment(node.nodeValue) != null) {
                return true
            }
        }
        return false
    }

    private fun parseSectionComment(raw: String?): String? {
        val value = raw?.trim() ?: return null
        if (!value.startsWith("#")) {
            return null
        }
        val label = value.removePrefix("#").trim()
        return if (label.isEmpty()) null else label
    }

    private fun addSection(rows: MutableList<TuixtAdapter.SettingsRow>, section: String) {
        if (rows.isNotEmpty() && rows.last().section == section) {
            return
        }
        rows.add(TuixtAdapter.SettingsRow.section(section))
    }

    private fun addGroupedRow(
        rows: LinkedHashMap<String, MutableList<XMLPrefsSave>>,
        section: String,
        save: XMLPrefsSave
    ) {
        var bucket = rows[section]
        if (bucket == null) {
            bucket = ArrayList<XMLPrefsSave>()
            rows[section] = bucket
        }
        bucket.add(save)
    }

    private fun flattenGroupedRows(
        groupedRows: LinkedHashMap<String, MutableList<XMLPrefsSave>>
    ): MutableList<TuixtAdapter.SettingsRow> {
        val rows: MutableList<TuixtAdapter.SettingsRow> = ArrayList()
        for (entry in groupedRows.entries) {
            val section = entry.key
            addSection(rows, section)
            for (save in entry.value) {
                rows.add(TuixtAdapter.SettingsRow.setting(save, section))
            }
        }
        return rows
    }

    @SuppressLint("GestureBackNavigation")
    override fun onBackPressed() {
        attemptClose()
    }

    private fun attemptClose() {
        if (!hasUnsavedChanges()) {
            setResult(BACK_PRESSED)
            finish()
            return
        }

        TuixtDialog.showConfirm(
            this,
            "Discard Changes?",
            "Unsaved settings changes will be lost.",
            "Discard",
            "Keep Editing",
            ConfirmAction {
                setResult(BACK_PRESSED)
                finish()
            })
    }

    private fun hasUnsavedChanges(): Boolean {
        if (adapter != null && adapter!!.hasPendingChanges()) {
            return true
        }
        if (plainTextEditor != null) {
            return !TextUtils.equals(originalRawText, plainTextEditor!!.getText().toString())
        }
        return false
    }

    companion object {
        const val PATH: String = "path"
        const val ERROR_KEY: String = "error"
        const val BACK_PRESSED: Int = 2
        const val SAVE_PRESSED: Int = 3
    }
}
